plenum/server/primary_selector.py

```diff
@@ -1,41 +1,284 @@
+
+    def _mark_replica_as_changed_view(self,
+                                      instance_id,
+                                      replica_name,
+                                      new_primary_replica_name,
+                                      ledger_info):
+        if instance_id not in self._view_change_done:
+            self._view_change_done[instance_id] = {}
+        if replica_name in self._view_change_done:
+            return False
+        data = (new_primary_replica_name, ledger_info)
+        self._view_change_done[instance_id][replica_name] = data
+        return True
+
+    def _hasViewChangeQuorum(self, instance_id):
+        # This method should just be present for master instance.
+        """
+        Checks whether 2f+1 nodes completed view change and whether one 
+        of them is the next primary
+        """
+        declarations = self._view_change_done.get(instance_id, {})
+        num_of_ready_nodes = len(declarations)
+        quorum = get_strong_quorum(f=self.f)
+        enough_nodes = num_of_ready_nodes >= quorum
+        if not enough_nodes:
+            return False
+
+        next_primary_name = self._who_is_the_next_primary(instance_id)
+        if next_primary_name not in declarations:
+            logger.debug("{} got enough ViewChangeDone messages "
+                         "for quorum ({}), but the next primary {} "
+                         "has not answered yet"
+                         .format(self.name,
+                                 num_of_ready_nodes,
+                                 next_primary_name,
+                                 instance_id))
+            return False
+
+        logger.info("{} got view change quorum ({} >= {}) for instance {}"
+                     .format(self.name,
+                             num_of_ready_nodes,
+                             quorum,
+                             instance_id))
+        return True
+
+    def _startSelection(self):
+        logger.debug("{} starting selection".format(self))
+        for instance_id, replica in enumerate(self.replicas):
+            if replica.primaryName is not None:
+                logger.debug('{} already has a primary'.format(replica))
+                continue
+            new_primary_name = self._who_is_the_next_primary(instance_id)
+            self._send_view_change_done_message(instance_id,
+                                                      new_primary_name,
+                                                      self.viewNo)
+            logger.display("{} selected primary {} for instance {} (view {})"
+                           .format(replica,
+                                   new_primary_name,
+                                   instance_id,
+                                   self.viewNo),
+                           extra={"cli": "ANNOUNCE",
+                                  "tags": ["node-election"]})
+
+            replica.primaryChanged(new_primary_name)
```

Comment:
Why do you think we need to notify replica and primary here? It complicates situation when we need to restart view change if other nodes selected another primary for some reason.